### ISSUE TITLE
More complex Example app

### ISSUE THREAD

**Issue by mxstbr** [styled-components/styled-components#1358]
**TL;DR: We need your help to build a React app covering all of the styled-components API to help us make sure the library works as expected!**

----

As of #1357 we screenshot integration test [our example app](https://github.com/styled-components/styled-components/tree/master/example) in CI for every PR, which is super awesome to make sure styled-components works as expected. Unfortunately, right now the example app isn't very complex, it's really only a couple lines of code.

We need a bigger Example app which leverages the entire styled-components public API and many CSS edge cases so we know that most things are working as expected. What specific content that app has doesn't really matter, that's all up to you. (massive bonus points if it looks great though!) 

The app lives in the `example/` folder in the repo. Whatever specific content you come up with for the app, it should use all of the following things at least once:

- styled-components API:
  - [ ] `injectGlobal`
  - [ ] `css`
  - [ ] `keyframes` (note: we screenshot the app in ci, so this has to be something that's applied immediately somehow and doesn't change over time?)
  - [ ] `styled.x`
  - [ ] `styled.x.attrs`
  - [ ] `styled(SomeComp)`
  - [ ] `SomeStyledComp.extend`
  - [ ] `SomeStyledComp.withComponent`
    - Theming
    - [ ] `ThemeProvider`
    - [ ] `withTheme`
  - Interpolations
    - [ ] `css` fragments
    - [ ] Static interpolations
    - [ ] Function interpolations with props
    - [ ] Interpolating other components
- CSS features
  - [ ] Media queries
  - [ ] Pseudo-selectors
  - [ ] Nesting
  - [ ] Parent selectors
- whatever else you can come up with!

Given the size of this task and its ambiguity it makes sense to have multiple people tackle this together. A great first step would be to come up with a quick plan for the contents of the app and then start coding! Any ideas?

**Comment by briandridge** [comment]
I'm going to start working on this, if anyone else wants to contribute let me know.

**Comment by GGAlanSmithee** [comment]
I'd love to contribute towards this. Probably will be a bit low on time over the holidays though. As for what the content should be, maybe simply display example code for each example, styled accordingly. Would make it easy for multiple authors to contribute towards this.

**Comment by briandridge** [comment]
Sounds great, that's what I was thinking too. Thinking of basically rendering multiple <Content> components in the App, one for each API/feature (some could probably be combined).
 
@mxstbr does this sound like what you are looking for?

**Comment by dreamorosi** [comment]
Hi everyone, huge fan of SC, I’d really love to contribute. Can we keep in touch to come up with a plan?

**Comment by mxstbr** [comment]
@briandridge that sounds perfect, it'd be awesome if we could even easily glance which feature was broken based on which part of the screenshot looks off.

Let's discuss this in the community? I've opened a new thread here: https://spectrum.chat/styled-components?thread=d475965b-2031-40c0-90e5-bb8159d46877 /cc @dreamorosi @briandridge @GGAlanSmithee

**Comment by maciej-ka** [comment]
ATM screenshot test is using project in `example` directory. Let's remove `sandbox` or `example` and keep just one of them?

**Comment by mxstbr** [comment]
> The project to create showcase is located in example dir

Ohh, I did not realize that, thanks for the correction!

**Comment by probablyup** [comment]
@mxstbr We should probably also use the sample app for benchmarking purposes

**Comment by GGAlanSmithee** [comment]
I made the OP into a checklist, so we can check of each item as we go.

**Comment by alanbsmith** [comment]
Hey! I don't know if you're looking for more help on this or not, but this is a basic starter I made: https://github.com/alanbsmith/styled-components-starter

Might be a good starting point? The main wins here are:
* setup for loading / configuring fonts from a CDN 
* setup for theme / ThemeProvider
* setup for injectGlobalStyles
* Jest test setup
* Style and ESLint setup

**Comment by alanbsmith** [comment]
The example page is just a series of buttons with the theme and font loaded:
http://recordit.co/xZS5Bvc41f

**Comment by mxstbr** [comment]
Seems to be a bit too much geared towards getting started with styled-components vs. an app that we can screenshot in CI?

**Comment by probablyup** [comment]
What about a TODO MVC situation?


On December 18, 2017 at 11:44:56 AM, Max Stoiber (notifications@github.com)
wrote:

Seems to be a bit too much geared towards getting started with
styled-components vs. an app that we can screenshot in CI?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/styled-components/styled-components/issues/1358#issuecomment-352483736>,
or mute the thread

**Comment by mxstbr** [comment]
I don't know if we really need any interaction? Realistically the only thing that app'll be used for is screenshots, but I don't particularly care tbch

**Comment by probablyup** [comment]
That's fair. I just figured it's a ready made app available in a bunch of different libraries that should be easy to restyle :)

**Comment by mxstbr** [comment]
That's true! Whatever y'all wanna do I'm open for, honestly!

**Comment by alanbsmith** [comment]
Gotcha. I misunderstood the purpose. Yeah, what I have here is geared toward helping people get started. But if this is just for CI, we can gut a lot of what's in there.

**Comment by dreamorosi** [comment]
So we were discussing this on Spectrum, as @mxstbr suggested, and we were kinda going toward doing very basic examples, like those found on the documentation pages, and then expand them where needed. As for the format we were thinking about mimicking the one used in the docs with the code blocks on the left and the rendered result on the right so we can keep things organised and know what is what.

I don't think we need anything too fancy and since it's gonna be used for screenshots only there's no point in making it interactive (?).

**Comment by GGAlanSmithee** [comment]
Agree with @dreamorosi

**Comment by GGAlanSmithee** [comment]
We are discussing this over at spectrum. I have one question though; since this is for screenshot testing, will this work with scrolling, alternatively multiple pages? @maciej-ka

**Comment by mxstbr** [comment]
We could probably do a scrolling screenshot, /cc @philpl

**Comment by lifeiscontent** [comment]
can someone describe how to use styled-theming when a component is composed of multiple elements? can someone update my example so I can see how you would apply multiple themes to an element like this? https://codesandbox.io/s/84qpnz9x42

**Comment by lifeiscontent** [comment]
@mxstbr why did you delete my comment?

**Comment by mxstbr** [comment]
@lifeiscontent I deleted your comment because it's totally irrelevant to the issue at hand and notified a bunch of people. Please ask questions like that on https://spectrum.chat/styled-components

**Comment by lifeiscontent** [comment]
ok, sorry.

**Comment by gretzky** [comment]
@mxstbr is there a status update on this? i might have a use case if this is still needed

**Comment by mxstbr** [comment]
Yeah this would still be needed I think!

**Comment by gretzky** [comment]
I am @abolajibisiriyu, but feel free to also work on it as I only have some time here and there to work on it so if you beat me to it, that's ok 😄

**Comment by lifeiscontent** [comment]
if anyone here is in San Francisco, I'd be down to meetup and pair on some examples together, otherwise if someone is down for some remote pairing, that could be an option too!

**Comment by lifeiscontent** [comment]
@mitchellhamilton / @mxstbr would you be available for a google hangout / discord voice call? I'd love to get your thoughts on a few things and plan some ideas out with you, first helping me on my project and then taking the knowledge and applying it to something like this.

**Comment by BarbaraPruz** [comment]
If needed, I'd love to help out on this.   @lifeiscontent - if you have some ideas on getting started and want some help, let me know.  Available for call or paired programming for some time most weekdays (Honolulu time)

**Comment by gretzky** [comment]
@lifeiscontent, do you want me to stop what i was doing then? Because above your comment, I noted I was doing some work on this already. @abolajibisiriyu also asked if they could do work on this as well. It would be nice if you could include everyone who has chimed in.

**Comment by lifeiscontent** [comment]
@gretzky I think we should all come up with some ideas, and maybe have an examples folder kind of like what next.js is doing.

**Comment by BarbaraPruz** [comment]
So something like 
- a basic example that could also cover passing props, and extending styles (including external lib), 
- an example using themes and related APIs
- an example using different pseudo selectors and pseudo elements
- and example for refs
- an example mixing in with straight CSS and demonstrating expected specificity behavior
- an example for animations (or work it into one of the other examples)
etc.

**Comment by Tolsee** [comment]
@mxstbr Is this issue stale. Or, we are open to contributing to this one? 

From what I can see, the integration test uses the `example` app. And there is not much in the example app. Can I add some more components to the example app, including more cases as you said in the comment above?

**Comment by maxxgl** [comment]
Since this issue seemed stale, I went ahead and started something a little more complex. I wasn't sure how much the /example app should be changed, so I started small to validate I was on the right track. Many of the components we're lifted from the documentation.

There are a plenty of improvements that can be made, both to the code quality, and the test concepts.

The new app can be seen here:
![example-test-js-example-page-should-match-screenshot-1-snap](https://user-images.githubusercontent.com/20944914/54865326-25566c80-4d64-11e9-8938-f3965a718933.png)

**Comment by ferdelamad** [comment]
@maxxgl awesome examples! Do you think I can contribute to this? I really want to start contributing to styled-components and I think this could get me started.
I have a couple of ideas that could be added (like using style objects) :
```
const Wrapper = styled.div(({ borderRadius, style }) => ({
  backgroundColor: 'lightblue',
  borderRadius: `${borderRadius || '25px'}`
  height: '100px',
  width: '100px'
  ...style,
}))
```
```
<Wrapper style={{ color: 'red', height: '200px' }} borderRadius={'20px'} />
```

**Comment by maxxgl** [comment]
I was finally able to get his to build in #2458. From the PR:

> The current Travis [snapshot test](https://github.com/maxxgl/styled-components/blob/1358-improve-example-app/packages/styled-components/integration-test/example.test.js) uses the `google-chrome-beta` binary as set in the Travis config (which I didn't realize at first). Which means that there are potentially quite a few problems when people test this locally, due to either the browser version being used by puppeeter, or the host system's browser quirks (especially font-related).
>
> Honestly, after getting a little frustrated trying to make this pass tests, I'm not sure I would recommend merging this considering the current snapshot strategy.

**Comment by maxxgl** [comment]
@ferdelamad Sorry for not getting back to you sooner, I kept thinking I was going to get this to pass, "just 20 more minutes" style. I think we should discuss if there is a more consistent way to test this before we put more work into it.

**Comment by kitten** [comment]
Closing due to age, inactivity, missing reproduction, or lack of details.
